ppci/lang/c/codegenerator.py

```diff
@@ -1451,39 +1458,31 @@ def gen_fill_varargs(self, var_args):
         """
         if var_args:
             # Allocate a memory slab:
-            size = 0
-            alignment = 1
-            for va in var_args:
-                va_size, va_alignment = self.data_layout(va.typ)
-                # If not aligned, make it happen:
-                size += required_padding(size, va_alignment)
-                size += va_size
-                alignment = max(alignment, va_alignment)
+            # each argument is stored in a slot, aligned on the strongest alignment
+            size = self.va_alignment * len(var_args)
+            alignment = self.va_alignment
```

Comment:
How about structs larger than the va_alignment size, is this allowed? Or can only basic types be passed via vararg way?

Reply:
There is no type restrictions in the C standard, although I have never seen composite types passed to vararg functions.
Struct/union are possible since they support argument-passing by value, and assignment.

In function calls; we can copy their value starting at one slot and occupying several slots. After the copy we realign the vararg pointer to the beginning of next slot.
In va_arg macro, we use assignment to get the value, we increment the vararg pointer by the type size + the required padding to realign the vararg pointer to the beginning of next slot.